File: packages/septr-go/env_check_test.go
package septr

import (
	"os"
	"path/filepath"
	"strings"
	"testing"
)

const (
	testKeyA = "septr_live_aaaaaaaa-aaaa-4aaa-8aaa-aaaaaaaaaaaa_ffffffffffffffffffffffffffffffff"
	testKeyB = "septr_live_bbbbbbbb-bbbb-4bbb-8bbb-bbbbbbbbbbbb_0123456789abcdef0123456789abcdef"
)

func TestMaskKey(t *testing.T) {
	if got := maskKey("short"); got != "***" {
		t.Errorf("expected *** for short key, got %s", got)
	}
	if got := maskKey(testKeyA); got != "septr_live_aaaa…aaaa_ff…ffff" {
		t.Errorf("unexpected mask: %s", got)
	}
	if got := maskKey(testKeyB); got != "septr_live_bbbb…bbbb_01…cdef" {
		t.Errorf("unexpected mask: %s", got)
	}
}

func TestReadDotenvKey(t *testing.T) {
	dir := t.TempDir()
	env := filepath.Join(dir, ".env")
	os.WriteFile(env, []byte("# comment\nSEPTR_API_KEY=\"septr_live_aaaa\"\nSEPTR_TELEMETRY_URL='http://x'\n"), 0o600)
	if got := readDotenvKey(env, "SEPTR_API_KEY"); got != "septr_live_aaaa" {
		t.Errorf("expected septr_live_aaaa, got %q", got)
	}
	if got := readDotenvKey(env, "SEPTR_TELEMETRY_URL"); got != "http://x" {
		t.Errorf("expected http://x, got %q", got)
	}
	if got := readDotenvKey(env, "MISSING"); got != "" {
		t.Errorf("expected empty for missing key, got %q", got)
	}
	if got := readDotenvKey(filepath.Join(dir, "nope.env"), "SEPTR_API_KEY"); got != "" {
		t.Errorf("expected empty for missing file, got %q", got)
	}
}

func TestEnvDotenvCandidates(t *testing.T) {
	dir := t.TempDir()
	os.MkdirAll(filepath.Join(dir, "sub"), 0o755)
	got := envDotenvCandidates(dir)
	if !hasPath(got, filepath.Join(dir, ".env")) {
		t.Errorf("missing cwd .env candidate: %v", got)
	}
	if !hasPath(got, filepath.Join(dir, "..", ".env")) {
		t.Errorf("missing parent .env candidate: %v", got)
	}
	if !hasPath(got, filepath.Join(dir, "sub", ".env")) {
		t.Errorf("missing child .env candidate: %v", got)
	}
}

func TestCheckEnvVsDotenv_WarnsOnce(t *testing.T) {
	dir := t.TempDir()
	env := filepath.Join(dir, ".env")
	os.WriteFile(env, []byte("SEPTR_API_KEY=\""+testKeyB+"\"\n"), 0o600)

	first := checkEnvVsDotenv(testKeyA, "SEPTR_API_KEY", []string{env})
	if len(first) != 1 {
		t.Fatalf("expected one warning, got %d", len(first))
	}
	if !strings.Contains(first[0], "bbbb…bbbb") || !strings.Contains(first[0], "aaaa…aaaa") {
		t.Errorf("warning missing masked keys: %s", first[0])
	}
	if strings.Contains(first[0], "ffffffffffffffffffffffffffffffff") {
		t.Errorf("warning leaked full secret: %s", first[0])
	}
	if second := checkEnvVsDotenv(testKeyA, "SEPTR_API_KEY", []string{env}); len(second) != 0 {
		t.Errorf("expected no second warning, got %v", second)
	}
}

func TestCheckEnvVsDotenv_MatchingKeySilent(t *testing.T) {
	dir := t.TempDir()
	env := filepath.Join(dir, ".env")
	os.WriteFile(env, []byte("SEPTR_API_KEY=\""+testKeyB+"\"\n"), 0o600)
	if got := checkEnvVsDotenv(testKeyB, "SEPTR_API_KEY", []string{env}); len(got) != 0 {
		t.Errorf("expected silence on match, got %v", got)
	}
}

func TestCheckEnvVsDotenv_NoEnvFileSilent(t *testing.T) {
	if got := checkEnvVsDotenv(testKeyA, "SEPTR_API_KEY", []string{filepath.Join(t.TempDir(), ".env")}); len(got) != 0 {
		t.Errorf("expected silence without .env, got %v", got)
	}
}

func TestCheckEnvVsDotenv_SilenceEnvVar(t *testing.T) {
	t.Setenv("SEPTR_SILENCE_ENV_WARNING", "1")
	dir := t.TempDir()
	env := filepath.Join(dir, ".env")
	os.WriteFile(env, []byte("SEPTR_API_KEY=\""+testKeyB+"\"\n"), 0o600)
	if got := checkEnvVsDotenv(testKeyA, "SEPTR_API_KEY", []string{env}); len(got) != 0 {
		t.Errorf("expected silence with SEPTR_SILENCE_ENV_WARNING, got %v", got)
	}
}

func hasPath(list []string, needle string) bool {
	for _, item := range list {
		if item == needle {
			return true
		}
	}
	return false
}

File: packages/septr-node/src/__tests__/env-check.test.ts
import { afterEach, describe, expect, it, vi } from "vitest"
import * as fs from "fs"
import * as path from "path"
import * as os from "os"
import {
  checkEnvVsDotenv,
  envDotenvCandidates,
  maskKey,
  readDotenvKey,
  resetEnvCheck,
  warnEnvVsDotenv,
} from "../core/env-check"

afterEach(() => {
  resetEnvCheck()
  delete process.env.SEPTR_SILENCE_ENV_WARNING
  vi.restoreAllMocks()
})

function tmpDir(): string {
  return fs.mkdtempSync(path.join(os.tmpdir(), "septr-env-check-"))
}

const KEY_A = "septr_live_aaaaaaaa-aaaa-4aaa-8aaa-aaaaaaaaaaaa_ffffffffffffffffffffffffffffffff"
const KEY_B = "septr_live_bbbbbbbb-bbbb-4bbb-8bbb-bbbbbbbbbbbb_0123456789abcdef0123456789abcdef"

describe("env-check", () => {
  it("masks keys keeping the project id visible", () => {
    expect(maskKey("short")).toBe("***")
    expect(maskKey(KEY_A)).toBe("septr_live_aaaa…aaaa_ff…ffff")
    expect(maskKey(KEY_B)).toBe("septr_live_bbbb…bbbb_01…cdef")
  })

  it("parses quoted and plain values from .env files", () => {
    const dir = tmpDir()
    const env = path.join(dir, ".env")
    fs.writeFileSync(env, '# comment\nSEPTR_API_KEY="septr_live_aaaa"\nSEPTR_TELEMETRY_URL=\'http://x\'\n')
    expect(readDotenvKey(env, "SEPTR_API_KEY")).toBe("septr_live_aaaa")
    expect(readDotenvKey(env, "SEPTR_TELEMETRY_URL")).toBe("http://x")
    expect(readDotenvKey(env, "MISSING")).toBeUndefined()
    expect(readDotenvKey(path.join(dir, "nope.env"), "SEPTR_API_KEY")).toBeUndefined()
  })

  it("discovers cwd, parent and child .env locations", () => {
    const dir = tmpDir()
    fs.mkdirSync(path.join(dir, "sub"))
    const candidates = envDotenvCandidates(dir)
    expect(candidates).toContain(path.join(dir, ".env"))
    expect(candidates).toContain(path.join(dir, "..", ".env"))
    expect(candidates).toContain(path.join(dir, "sub", ".env"))
  })

  it("warns once when env key differs from the local .env key", () => {
    const dir = tmpDir()
    const env = path.join(dir, ".env")
    fs.writeFileSync(env, `SEPTR_API_KEY="${KEY_B}"\n`)
    const first = checkEnvVsDotenv(KEY_A, "SEPTR_API_KEY", [env])
    expect(first).toHaveLength(1)
    expect(first[0]).toContain("bbbb…bbbb")
    expect(first[0]).toContain("aaaa…aaaa")
    expect(first[0]).not.toContain("ffffffffffffffffffffffffffffffff")
    expect(checkEnvVsDotenv(KEY_A, "SEPTR_API_KEY", [env])).toEqual([])
  })

  it("is silent when the env key matches the local .env key", () => {
    const dir = tmpDir()
    const env = path.join(dir, ".env")
    fs.writeFileSync(env, `SEPTR_API_KEY="${KEY_B}"\n`)
    expect(checkEnvVsDotenv(KEY_B, "SEPTR_API_KEY", [env])).toEqual([])
  })

  it("is silent when no .env file exists", () => {
    expect(checkEnvVsDotenv(KEY_A, "SEPTR_API_KEY", [path.join(tmpDir(), ".env")])).toEqual([])
  })

  it("respects SEPTR_SILENCE_ENV_WARNING", () => {
    process.env.SEPTR_SILENCE_ENV_WARNING = "1"
    const dir = tmpDir()
    const env = path.join(dir, ".env")
    fs.writeFileSync(env, `SEPTR_API_KEY="${KEY_B}"\n`)
    expect(checkEnvVsDotenv(KEY_A, "SEPTR_API_KEY", [env])).toEqual([])
  })

  it("prints the warning to stderr via warnEnvVsDotenv", () => {
    const dir = tmpDir()
    const env = path.join(dir, ".env")
    fs.writeFileSync(env, `SEPTR_API_KEY="${KEY_B}"\n`)
    const errorSpy = vi.spyOn(console, "error").mockImplementation(() => {})
    warnEnvVsDotenv(KEY_A, [env])
    expect(errorSpy).toHaveBeenCalledOnce()
    expect(String(errorSpy.mock.calls[0][0])).toContain("[septr] WARNING")
  })
})
